apps/manager_app/joystick_training.c

```diff
@@ -0,0 +1,254 @@
+ * all the early exit cases.
+ *
+ * @return Whether the check is pass or fail.
+ */
+static bool validate_training_request(manager_query_t *query);
+
+/**
+ * @brief Fill the error in the structure generated via protobuf.
+ * @details The function sets the necessary fields for the data related errors.
+ */
+static void send_training_error(uint32_t error_code);
+
+/**
+ * @brief Sends the encoded message via usb api to the host.
+ * @details The functions takes the value of fields from the provided reference
+ * to structure instance and encodes into a buffer before using the usb_send_msg
+ * to send data in appropriate format to the host app.
+ *
+ * @param training Reference to a filled structure instance of
+ * manager_train_joystick_response_t
+ */
+static void send_msg_to_host(manager_train_joystick_response_t *training);
+
+/**
+ * @brief Used to handle each joystick training
+ * @details The function will render display with received message and then
+ * listen for UI event for the expected user action on joystick.
+ * The function **ignores timeout events** since currently, there is not
+ * way to disable it. Also, in case of abort, it will clear usb event
+ * (usb_clear_event) and silently return.
+ * Upon receiving the UI event, the function will then update the core flow
+ * status.
+ *
+ * @param ctx Reference to the initialized instance of training_context_t
+ * @param step_index The current index of joystick training to be fetched from
+ * the provided context
+ *
+ * @return Returns if the step completed successfully.
+ * @retval true If the joystick training step succeeds.
+ * @retval false If any abort event is received
+ */
+static bool training_step(const joystick_step_t *step);
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static bool validate_training_request(manager_query_t *query) {
+  // verify host query data is correct
+  manager_train_joystick_request_t *req = &query->train_joystick;
+  if (MANAGER_QUERY_TRAIN_JOYSTICK_TAG != query->which_request ||
+      MANAGER_TRAIN_JOYSTICK_REQUEST_INITIATE_TAG != req->which_request) {
+    send_training_error(1);    // TODO: use correct data-error code
+    return false;
+  }
+
+  // TODO: check the request with what device knows
+  return true;
+}
+
+static void send_training_error(uint32_t error_code) {
+  manager_train_joystick_response_t training =
+      MANAGER_TRAIN_JOYSTICK_RESPONSE_INIT_ZERO;
+  training.which_response = MANAGER_TRAIN_JOYSTICK_RESPONSE_COMMON_ERROR_TAG;
+  training.common_error.which_error =
+      ERROR_COMMON_ERROR_DEVICE_SETUP_REQUIRED_TAG;
+  training.common_error.device_setup_required = error_code;
+  send_msg_to_host(&training);
+}
+
+static void send_msg_to_host(manager_train_joystick_response_t *training) {
+  uint8_t payload[MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE] = {0};
```

Comment:
might wanna update the macro here

Reply:
i will add a todo to replace this with relevant macro when gets defined.